Favor unique ids over hashcodes

krmahadevan opened the pull request:
Closes #2566

Fixes #2566  .

### Did you remember to?

- [x] Add test case(s)
- [x] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

juherr commented on core/src/main/java/org/testng/ITestResult.java:
```diff
@@ -108,6 +108,14 @@
+   * specific test method's result.
+   */
+  default String id() {
+    return null;
```
I think throwing an exception with a comprehensive message is better than an NPE later in `Reporter`.

krmahadevan replied:
@juherr - Are you suggesting that someone outside of TestNG is going to try and implement the `ITestResult` interface and that would somehow land up with the TestNG codebase ? I am not sure how that can happen. Can you please help me understand the use-case when that would happen ?

juherr replied:
> I am not sure how that can happen.

Agree! No default method, just a declaration then ;)

krmahadevan replied:
Fixed. I have also rebased off of master and fixed merge conflicts and pushed in the amended commit. Please help take a look.

krmahadevan commented:
@juherr - Please let me know if this PR can be merged.

vlsi commented:
Would it better use `IdentityHashMap`?

krmahadevan commented:
@vlsi  - That would cause the key to be `ITestResult` no ? If that's the cause, then I think this would increase the memory foot print as well right ?

vlsi commented:
What is the purpose of the `Reporter` then?

I guess one must have `ITestResult` in hand, how do they otherwise call `Reporter # public static synchronized List<String> getOutput(ITestResult tr) {` API?

If the idea is to throw away unclaimed outputs, then `WeakHashMap` (or its concurrent alternative or even storing the output right inside `ITestResult`) might be a better fit.

vlsi commented:
For instance, `ITestResult` implements `IAttributes`. What if the output was one of the attributes added by TestNG?

krmahadevan commented:
@vlsi 

> What is the purpose of the Reporter then? I guess one must have ITestResult in hand, how do they otherwise call Reporter # public static synchronized List<String> getOutput(ITestResult tr) { API?

The `ITestResult` objects are visible via listeners no ? So a user would be invoking it from there. The `ITestResult` already is bloated and there are long standing discussions that discuss about reducing the memory foot print. So I am not too keen on adding more to the memory consumption than what already exists in TestNG.

> If the idea is to throw away unclaimed outputs, then WeakHashMap (or its concurrent alternative or even storing the output right inside ITestResult) might be a better fit.

The output needs to be available for the user till the end of execution (including reporting phase). So a `WeakHashMap` would baiscally stand a chance of getting GC'ed somewhere in between no ? if that's the case, then the user stands to lose test logs, which is breakage of functionality.

krmahadevan commented:
> For instance, `ITestResult` implements `IAttributes`. What if the output was one of the attributes added by TestNG?

I didnt understand what you are referring to. Can you please elaborate? `IAttributes` serves a different purpose than the `ITestResult` object, but both of them are needed by a test result.

vlsi commented:
>I didnt understand what you are referring to.

Here's the current code:

https://github.com/cbeust/testng/blob/9733de34e4becabd2cc3f9434c0eb36d85135b70/core/src/main/java/org/testng/ITestResult.java#L14

What you have now is 

```java
class Reporter {
    private static final Map<String, List<Integer>> m_methodOutputMap = Maps.newHashMap();
    private static synchronized void log(String s, ITestResult m) {
        List<Integer> lines =
            m_methodOutputMap.computeIfAbsent(m.id(), k -> Lists.newArrayList());
```

What you could have is 

```java
class Reporter {
    private static final String TEST_OUTPUT_ATTRIBUTE = "Reporter.testOutput";
    private static synchronized void log(String s, ITestResult m) {
        List<Integer> lines = (List<Integer>) m.getAttribute(TEST_OUTPUT_ATTRIBUTE);
        ...
```

krmahadevan commented:
@vlsi - The attributes are available to the user as well. So how do I ensure that the user doesn't interfere with the test logging capability that TestNG provides ? 
On a different note, just trying to understand as to what issue do you see with the current UUID backed implementation as key instead of the earlier `hashCode()` ?

vlsi commented:
> The attributes are available to the user as well

Is it that bad?

> So how do I ensure that the user doesn't interfere with the test logging capability that TestNG provides ?

For instance, claiming a namespace would do. If all testng-added attributes start with `org.testng`, then the probability of collision would be virtually zero.

---

> what issue do you see with the current UUID backed implementation

You just added a public API method `String id()`. I am not sure that was planned. I know JUnit 5 has a complicated ID (test refers to its parent or something like that)

krmahadevan commented:
> Is it that bad?

Yes it is. Because now we are banking on the fact that the user will stay clear from the TestNG specific attribute keys which IMO is not very intuitive.

> For instance, claiming a namespace would do. If all testng-added attributes start with org.testng, then the probability of collision would be virtually zero.

Sure it can be done but it still doesnt stop someone from accidently corrupting the name space and there's no way to prevent it. 

> You just added a public API method String id(). I am not sure that was planned. I know JUnit 5 has a complicated ID (test refers to its parent or something like that)

So what is the issue with adding an API to an `ITestResult` interface which is only used by TestNG ? I can understand the impact if we end up adding an API which is meant for external user consumption, but ITestResult is meant to be implemented only by TestNG and it provides no support for a user to embed their implementations into TestNG. So not quite sure as to what the concern is here.

vlsi commented:
> So what is the issue with adding an API to an `ITestResult` interface which is only used by TestNG ? 

I am afraid `org.testng.ITestResult` is used a LOT in client-facing code.
For instance, `ITestListener` has 6 methods that receive `ITestResult` parameter (e.g. `onTestStart`, `onTestSuccess`)

krmahadevan commented:
> I am afraid org.testng.ITestResult is used a LOT in client-facing code.
For instance, ITestListener has 6 methods that receive ITestResult parameter (e.g. onTestStart, onTestSuccess)

Yes, but everywhere `ITestResult` is only consumed and no one implements the `ITestResult` interface. The impact of adding a new api into the interface contract without a default implementation will break client contracts, but its not the case here right.

vlsi commented:
>  The impact of adding a new api into the interface contract without a default implementation will break client contracts

1. They might have code that mocks `ITestResult` or something like that
1. They might start calling `id()` method and they might rely on it

vlsi commented:
If the output was attached to the test result itself, then it would make sense for the consumers as well. `output` does sound like `test result` for me, and it would be nice to see it in the debugger.

Package-level separation is very well known in Java, and it is understood in the community.

Here's a similar case from Gradle: https://docs.gradle.org/current/userguide/variant_attributes.html
They declare `well-known` names like `org.gradle.category`. The clients can either use string names, or they can use Java constants like `Category.CATEGORY_ATTRIBUTE`.

juherr commented:
@vlsi I agree but the risk is limited because most of the users won't do it and I don't see a real reason to do it, even for power users.

vlsi commented:
@juherr , there are cases when they implement `org.testng.ITestResult`
I am not sure why do they do that, however, they do that, and the added method would break their code even in case they do not use reporter API.

https://github.com/search?l=Java&p=2&q=%22implements+ITestResult%22&type=Code

https://github.com/flying0dancing/cabrite/blob/daf8f149b332e2adec4f307787faf132e2d5ef19/cabrite-core/src/main/java/com/csdental/test/Result.java#L6-L10

https://github.com/dongwenke/testng/blob/0e751441e53687fdc4666665f6febd6454c04b5b/src/test/java/test/reflect/TestResultJustForTesting.java#L6-L15

https://github.com/assertj/assertj-swing/blob/9dae1fa4857fb9fc02ed77bf72f12b3ded75c8d2/assertj-swing-testng/src/test/java/org/assertj/swing/testng/listener/TestResultStub.java#L18-L25

juherr commented:
😭 
So what could be the best option? 

* Removing the new method and only use it for specific implementation
* Use a default implementation (hashCode? throw an exception?)

krmahadevan commented:
@vlsi  - A quick check revealed that only the interface name matches, but they dont seem to be from TestNG.

Only https://github.com/assertj/assertj-swing/blob/9dae1fa4857fb9fc02ed77bf72f12b3ded75c8d2/assertj-swing-testng/src/test/java/org/assertj/swing/testng/listener/TestResultStub.java#L18-L25

is referring to the TestNG `ITestResult` and 

https://github.com/dongwenke/testng/blob/0e751441e53687fdc4666665f6febd6454c04b5b/src/test/java/test/reflect/TestResultJustForTesting.java#L6-L15 is a TestNG fork.

@juherr - I think we should just add back `id()` as a default implementation so that the anyone else who is implementing it wont face compilation problems. 

But I am still not able to understand as to what would they do with this implementation when TestNG doesn't provide any directly exposed api to replace the built in TestResult object with a user created TestResult object (without any hacks)

vlsi commented:
> So what could be the best option?

The best option is to use `ITestResult#setAttribute` to associate `ITestResult` with its output lines (it could be the current `List<Integer>` or `List<String>` or a custom interface). It solves the issue, it would automatically be garbage collected, and it requires no changes to the API. Note: memory-wise, `List<Integer>` and `List<String>` are the same.

krmahadevan commented:
> > So what could be the best option?
> 
> The best option is to use `ITestResult#setAttribute` to associate `ITestResult` with its output lines (it could be the current `List<Integer>` or `List<String>` or a custom interface). It solves the issue, it would automatically be garbage collected, and it requires no changes to the API. Note: memory-wise, `List<Integer>` and `List<String>` are the same.

@vlsi - For the record, I am not in favor of this since a test attribute is not the same as test logs. So this would be diluting the intent.

@juherr - I leave the decision to you :)

krmahadevan commented:
@vlsi 

And oh btw 

> It solves the issue, it would automatically be garbage collected, and it requires no changes to the API.

This is not true. There wont be a GC because till the end, an `ITestResult` object stays in memory. So there's no scope of a GC freeing up anything here.

vlsi commented:
>This is not true. There wont be a GC because till the end, an ITestResult object stays in memory. So there's no scope of a GC freeing up anything here.

@krmahadevan , 
1. if `Map<String, List<Integer>>` is kept as `static` field in `Reporter`, then its lifetime is not associated with `ITestResult`, and it might be that the `Lists<...>` are kept in memory even after `ITestResult` is reclaimed.
1. I already suggested `IdentityHashMap` and you claimed it would retain more memory due to `ITestResult` being referenced and kept in memory: https://github.com/cbeust/testng/pull/2571#issuecomment-852093748

I do not understand when `ITestResult` objects are created and removed, however, I claim that storing `output` inside `ITestResult` makes it way easier to reason about the lifetime than storing the same thing in `static field`.

krmahadevan commented:
@vlsi 

> I do not understand when ITestResult objects are created and removed, however, I claim that storing output inside ITestResult makes it way easier to reason about the lifetime than storing the same thing in static field.

I am not contesting the fact that the current implementation can do a bit of refactoring. But that refactoring cannot be at the cost of adding a different semantics to attribute is all that I am saying. If we are going to look at enhancing `ITestResult` by making it aware of `Logs` (this would be an enhancement to the API) that would be fine too. But just because we are worried about breaking backward compatibility to a publicly exposed API which does not have any semantic meaning outside of TestNG and whose instantiation and wiring in mechanism is strictly controlled by TestNG, then I feel that we are just doing a wild goose chase. I say this based on all these years of working with the TestNG codebase, wherein we have added APIs to the `ITestResult` object without any users complaining about it.

`ITestResult` already stays throughout the liftetime of TestNG execution. My only concern was burdening it with additional object information because for every execution `ITestResult` is absolutely needed, but not every execution logs data to the current test result.

This is my point of contention.

vlsi commented:
> But that refactoring cannot be at the cost of adding a different semantics to attribute is all that I am saying

`IdentityHashMap` would resolve "hashcode collision" issue without changing the APIs, and it would not add new attributes to `ITestResult`.
That was exactly what I suggested in the first comment: https://github.com/cbeust/testng/pull/2571#issuecomment-852091603

> we are worried about breaking backward compatibility to a publicly exposed API

The interface is public, and it is used a lot in client-facing code. Please refrain from making modifications to the interfaces like that.

Of course, if there's `org.testng.internal` interface, you are free to alter it in any way. However, if you have something public, then why alter it?

krmahadevan commented:
@vlsi 

> The interface is public, and it is used a lot in client-facing code. Please refrain from making modifications to the interfaces like that.

We seem to be going around in circles. Not all public APIs exposed by TestNG have a value beyond the API and cannot be implemented and plugged into TestNG (not atleast without a lot of hacking). So I dont agree to this rationale although from a general coding perspective, this is a good approach and should be followed as much as possible.

> However, if you have something public, then why alter it?

If there are no implementations possible that can be wired into TestNG, then why not ? That is what I have been saying all along. 

Beyond the TestNG listener interfaces which an end-user will actually implement, there are a lot of interfaces within TestNG which have gotten inadvertently exposed (they should have been `internal` but they aren't), but by behavior they are `internal` interfaces.

Please feel free to skim through the codebase and you will find many such interfaces. Applying a blanket rule to these interfaces IMO is not going to lead us anywhere.

vlsi commented:
@krmahadevan , please check the very beginning of `IdentityHashMap` javadoc:

>This class implements the Map interface with a **hash table**, **using reference-equality** in place of object-equality when **comparing keys** (and values). In other words, **in an IdentityHashMap**, two keys k1 and k2 are considered equal **if and only if** (k1==k2). (In normal Map implementations (like HashMap) two keys k1 and k2 are considered equal if and only if (k1==null ? k2==null : k1.equals(k2)).)

And, yes, it works even in case the default `Object#hashCode` returns 1 for all the instances. Of course, the map becomes slower, however, it still works.

krmahadevan commented:
Thanks for clarifying that part @vlsi

juherr commented:
> @juherr - I leave the decision to you :)

Someone could summarize the different options?
